collectors/proc.plugin/proc_spl_kstat_zfs.c

```diff
@@ -194,3 +198,206 @@ int do_proc_spl_kstat_zfs_arcstats(int update_every, usec_t dt) {
+{
+    (void)dt;
+
+    static int do_zfs_pool_state = -1;
+    static char *dirname = NULL;
+
+    int pool_found = 0, state_file_found = 0;
+
+    if (unlikely(do_zfs_pool_state == -1)) {
+        char filename[FILENAME_MAX + 1];
+        snprintfz(filename, FILENAME_MAX, "%s%s", netdata_configured_host_prefix, "/proc/spl/kstat/zfs");
+        dirname = config_get("plugin:proc:" ZFS_PROC_ARCSTATS, "directory to monitor", filename);
+
+        zfs_pools = dictionary_create(DICTIONARY_FLAG_SINGLE_THREADED);
+
+        do_zfs_pool_state = 1;
+    }
+
+    if (likely(do_zfs_pool_state)) {
+        DIR *dir = opendir(dirname);
+        if (unlikely(!dir)) {
+            error("Cannot read directory '%s'", dirname);
+            return 1;
+        }
+
+        struct dirent *de = NULL;
+        while (likely(de = readdir(dir))) {
+            if (likely(
+                    de->d_type == DT_DIR && ((de->d_name[0] == '.' && de->d_name[1] == '\0') ||
+                                             (de->d_name[0] == '.' && de->d_name[1] == '.' && de->d_name[2] == '\0'))))
+                continue;
+
+            if (unlikely(de->d_type == DT_LNK || de->d_type == DT_DIR)) {
+                pool_found = 1;
+
+                struct zfs_pool *pool = dictionary_get(zfs_pools, de->d_name);
+
+                if (unlikely(!pool)) {
+                    struct zfs_pool new_zfs_pool = {};
+                    pool = dictionary_set(zfs_pools, de->d_name, &new_zfs_pool, sizeof(struct zfs_pool));
+                };
+
+                pool->updated = 1;
+
+                if (pool->disabled)
+                    continue;
+
+                char filename[FILENAME_MAX + 1];
+                snprintfz(
+                    filename, FILENAME_MAX, "%s%s/%s/state", netdata_configured_host_prefix, ZFS_PROC_POOLS, de->d_name);
+
+                char state[STATE_SIZE + 1];
+                int ret = read_file(filename, state, STATE_SIZE);
+
+                if (!ret) {
+                    state_file_found = 1;
+
+                    // ZFS pool states are described at https://docs.oracle.com/cd/E19253-01/819-5461/gamno/index.html
+                    if (!strcmp(state, "ONLINE\n")) {
+                        pool->online = 1;
+                    } else if (!strcmp(state, "DEGRADED\n")) {
+                        pool->degraded = 1;
+                    } else if (!strcmp(state, "FAULTED\n")) {
+                        pool->faulted = 1;
+                    } else if (!strcmp(state, "OFFLINE\n")) {
+                        pool->offline = 1;
+                    } else if (!strcmp(state, "REMOVED\n")) {
+                        pool->removed = 1;
+                    } else if (!strcmp(state, "UNAVAIL\n")) {
+                        pool->unavail = 1;
+                    } else {
+                        disable_zfs_pool_state(pool);
+                        error("ZFS POOLS: Undefined state %s for zpool %s, disabling the chart", state, de->d_name);
```

Comment:
The `state` value contains newline character (\n). I think we want to remove it when logging an undefined state?

Reply:
Fixed.